### DISCUSSION TITLE
Methodology for solving computational homogenization problems

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#19997] [Q&A General] [answered]
Dear all,
I am finally down to the problem of selecting the right approach for performing homogenization calculations for my work.  (https://github.com/idaholab/moose/discussions/18776) can be referred for the background information related to this work. The sub-model that is currently in use is driven by a stress tensor and solved by a global strain approach. I initially found this to be suitable to my requiremements when things are raw and not matured. As i tend to refine my approach to the solution of this multi-level FE problem i found out that the sub-model i am executing is supposed to be driven by a macroscopic deformation gradient, rather than a conventional stress tensor. The attached paper can be referred back to understand the specifics involved in this approach.
[Geers_computational_homogenization.pdf](https://github.com/idaholab/moose/files/7840936/Geers_computational_homogenization.pdf)
As you can refer from `section 4.2` the micro-scale problem is mainly influenced by the macro deformation gradient for each gauss point, that plays a key role in the macro-micro kinematcis and the ensuing boundary conditions. Please could you suggest if there are any such alternate methods available in MOOSE that can execute an RVE model purely through the application of macro deformation gradient? Also i will need to homogenize the PK stress tensor and tangent stiffness matrix and so i want the sub-model to be operating under a finite strain space. Can this be a possibility?
Kind regards,
Arun

**Comment by GiudGiud** [comment]
@SudiptaBiswas @sapitts

**Comment by jiangwen84** [comment, marked as answer]
It is possible but I do not think it is even been solved before yet. For this, you need a multi-app models in which the master app is the macro problem and sub apps are the micro problem. If my understanding is correct, at each quadrature point of the macro problem, you need to solve a micro problem. For that you need to transfer the information, such as macro deformation gradient and homogenized PK stress and tangential stiffness from/to master to/from sub apps. That means you need implement customized `transfer` and figure out how to calculate/store those data.

**Reply by abarun22** [reply]
Thanks Wen. Thats indeed what i am working at the moment. I did create a method earlier for the customized transfer of key data items you listed here. I am currently trying to check how i can set up the RVE model, essentially the periodic boundary conditions driven by the macro deformation gradient tensor F<sub>M</sub>. Moose currently can handle the general PBC's by ensuring equal displacements along the periodic boundaries, u<sup>+</sup> = u<sup>-</sup>.
For an F<sub>M</sub> driven RVE you may want to say (F<sub>M</sub> - I) (X<sup>+</sup> - X<sup>-</sup>) = u<sup>+</sup> - u<sup>-</sup>. Here `I` is the identity tensor and X being the position vector in the dimensional space. 
Re-writing this equation for the repsective periodic faces we can say
(F<sub>M</sub> - I) X<sup>+</sup> - u<sup>+</sup> = (F<sub>M</sub> - I) X<sup>-</sup> - u<sup>-</sup>

The LHS and RHS literally implies the displacement quantities for the opposite periodic faces and this would seem analogous to U<sup>+</sup> = U<sup>-</sup> and eventually required to be implemented in to PBC's. My intention here is to check if there are any possibilities to represent this expression ((F<sub>M</sub> - I) X- u) as PBC constraints through existing features available in MOOSE. Probably somebody might've tried similar attempts while addressing computational homogenization problems. Any thoughtful ideas here would be much appreciated.
Kind regards,
Arun

**Reply by jiangwen84** [reply]
MOOSE uses `GlobalStrain` approach to enforce periodic BCs for mechanics problems. See https://mooseframework.inl.gov/syntax/Modules/TensorMechanics/GlobalStrain/

**Reply by abarun22** [reply]
Hi Wen,
Thanks for the suggestions. More questions from my side.

1. It seems like the global strain approach is another way of enforcing periodicity constraints while running an RVE sub-model, however driven by a macro stress tensor. I wonder whether this model gives equivalent results as that of the conventional RVE models for computational homogenization driven by a macro deformation gradient. I was particularly keen to know how the micro-fluctuation field is super-imposed in the calculation of current configuration in MOOSE as it was the case for the F<sub>M</sub> driven analysis 
given here [Geers_Kuznetsova_EncyclopCompMech_Chapter.pdf](https://github.com/idaholab/moose/files/7925190/Geers_Kuznetsova_EncyclopCompMech_Chapter.pdf) in section 4.2.
2. I would want to see the global strain approach to be extended to large strain material models, so that calculation of PK1 (for homogenization) makes relevance there. I am not sure how extensive this work is, if one wants to take initiative and do the extension activities.
3. Is it necessary to use CPFE material models alongside global strain models to be able to generate the right results reflecting the micro material-behavour at this scale. At the current level of implementation CPFE models aren't supported and so additional work required here.
4. Do you see the homogenized tangent stiffness matrix to be returned to the macro model along with PK1?
Kind regards,
Arun

**Reply by GiudGiud** [reply]
@jiangwen84

**Reply by SudiptaBiswas** [reply]
We have a publication, where you can find more about the applicability of the `GlobalStrain` approach.  It has mostly been used for RVE based mesoscale simulations. https://www.sciencedirect.com/science/article/pii/S0168874X20301165

**Reply by abarun22** [reply]
Hi Sudipta,
I find GlobalStrain approach to be suitable for solving RVE homogenization problems but with certain adjustments. Once after the RVE solution phase is completed i may need to add up the macro-displacement field with the fluctuation field to arrive at the current configuration. The macro-deformation gradient is further added up with the deformation gradient due to fluctuation field in order to obtain the micro-deformation gradient. This would then help me to compute the Piola stresses needed to be scaled to the macro gauss points. I would be required to work in a large strain setting if want to compute these stress measures and accomplish the task of homogenization. May i ask your help in porting the global strain models to work under an incremental strain setting [#18776](https://github.com/idaholab/moose/discussions/18776), so that we provide an interface to access large strain material models. Without this activity we may not be able to achieve the correct set of stress measures for homogenization. 
Thanks and regards,
Arun

**Reply by SudiptaBiswas** [reply]
Hi Arun, Unfortunately, I won't have the opportunity to work on it any time soon.

**Reply by reverendbedford** [reply]
This was done as part of the new tensor_mechanics "Lagrangian" kernels.  See https://mooseframework.inl.gov/modules/tensor_mechanics/Homogenization.html.  You can drive the RVE with either an imposed 1st PK stress tensor, an imposed deformation gradient, or a combination of both.

**Reply by abarun22** [reply]
Thanks for suggesting this method. It indeed looks very much interesting. Before integrating this RVE model, i wanted to clarify certain things, mainly on the front of data quantities that needed to be passed between the macro and RVE models to accomplish the workflow. Looking at the system documentation, it seems like this method primarily aims at imposing scale transition constraints based on the targets set by the user. These targets are either the stress or deformation gradient measures for a model that exhibits a large strain kinematic behaviour. My question is how do we know these targets before we intend to run the RVE model. Essentially the effective properties (PK1) that we might expect to compute from a typical homogenization scheme, are volume averaged and are equivalent to these targets which are eventually transferred to the macro gauss points. This scenario now presents with a dilemma as to what quantities are to be passed from and to the macro gauss points. Any clarification on this would be highly appreciated.

**Reply by reverendbedford** [reply]
The imposed deformation gradient on the microscale problems would come from the displacement gradient in the macroscale problem.  The stress contributing to the residual in the macroscale problem would be the averaged PK1 stress calculated via the homogenization system over the microscale problems.

If you're asking about how to connect the microscale and macroscale problems I haven't done anything like that in the past.

**Reply by abarun22** [reply]
Hi Mark,
When the macro-scale displacement gradients are imposed on the micro-scale problems we obtain an averaged PK1 that is 
1) Compatible with the input displacement gradients and 
2) Consistent with the other case where we input only stresses to drive the micro problem
In other words this means that we get similar results whether we run the micro problem with stresses or diplacement gradient alone. This suggests that the homogenizaiton system just reflects what has been input in the first place, since it solves constrained equation oncourse to computing the actual micro stresses. If we run the micro-problem with stresses alone then we may as well end up sending the same stresses to the macro gauss points even after the homogenization calculations, which certainly is incorrect. I am not very sure whether the applied displacement gradients produces homogenized results which are compatible to be plugged in to the respective macro gauss points.

**Reply by reverendbedford** [reply]
If you do things this way and iterate between the two scales as needed to solve for stress equilibrium on the macroscale the displacements in the microscale problems will be compatible with the macroscale deformation (by construction) and the stresses will be equilibrated in the weak sense.

**Reply by abarun22** [reply]
Hi Mark,
Yes, i agree with you that stress equilibrium of the macro model will be maintained in a weaker sense after mapping of homogenized stress measures. I was initially concerned how different these homogenized stresses are as against the original macro stresses, if we are in for a displacement gradient driven approach. If these are different, realistic and representative of the behaviour influenced by the micro model, then we are in a good shape proceeding further with this approach. Also i had a look at the code to get a first hand idea of the implementation behind the construction of residual. I could not locate the place where the volume averaging of the deformation gradient is carried out. I see the following lines in `HomogenizationConstraintIntegral.C`.

```
else if (_ctypes[_h] == HomogenizationConstants::ConstraintType::Strain)
      {
        Real f = (_indices[_h].first == _indices[_h].second) ? 1.0 : 0.0;
        res(_indices[_h].first, _indices[_h].second) =
            _F[_qp](_indices[_h].first, _indices[_h].second) -
            (f + _targets[_h]->value(_t, _q_point[_qp]));
      }

```
Here `_F` is the quantity we are expecting, possibly from the material modules, and used in the residual calculation. Am i correct in assuming that this quantity is already in a form that is volume averaged, to be used directly in the calculation of `res`. Essentially i am looking for the quantity (averaged PK1) that is needed to be plugged in to corresponding macro gauss points.
Kind regards,
Arun

**Reply by reverendbedford** [reply]
You would need to modify the existing structure somewhat.  

1. Replace the [function](https://github.com/idaholab/moose/blob/bc2965deaa4a247b0f52d4f3a52487507c92037a/modules/tensor_mechanics/src/userobjects/lagrangian/HomogenizationConstraintIntegral.C#L34) targets with your coupled deformation gradient from the macroscale problem. 
2. You can get the volume-averaged PK1 components with a [postprocessor](https://mooseframework.inl.gov/source/postprocessors/ElementAverageValue.html).  The material property you want is `"pk1_stress"`.
3. One thing I didn't think about: you'll want the derivative of the cell average PK1 stress with respect to the macroscale deformation gradient to provide the Jacobian in the macroscale problem.  Calculating this guy is some sort of integral over the domain just like what happens [here](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/userobjects/lagrangian/HomogenizationConstraintIntegral.C) but you'll need to work out the math for this yourself.  The integral will involve the material property `"pk1_jacobian"`.

**Reply by abarun22** [reply]
Hi Mark,
As you said we may need some enhancements to be able to suit the requirements we are working towards. I do not understand the need for derivative of the cell average PK1 stress with respect to the macroscale deformation gradient. Does the macro-model needs this quantity (jacobian) to be referenced at any stage during the computation? Do you see this derivative to be going inside an another user object that will be developed as part of the enhancement? Another thing i wanted to ask you is on the calculation of residual - how does the system ensures the residual that should go either to zero or a negligible number? Do you see any iterative calculations involved here that forces the volume averaged `_F` to match with the equivalent target value?
Kind regards,
Arun

**Comment by abarun22** [comment]
Hi Mark,
I started working towards integrating this system in to my workflow. I am keen to know the requirement for item 3 in your previous note. Please could you let me know?

**Reply by reverendbedford** [reply]
Sorry, I missed the notification you replied.

1. You need that derivative in the macroscale problem to solve for equilibrium.
2. Presumably you would calculate it as a scalar AuxVariable in your microscale problems with some sort of user object and then transfer it to the material property ("pk1_jacobian") in the macroscale problem.
3. The usual MOOSE way.  You can either use newton's method or PJFNK.  If you are careful with your Jacobians I would use newton.
4. Yes, there will be iterations on both the macroscale problem (equilibrium of the macroscale stress field) and on the microscale problems (equilibrium on the microscale + complying with the deformation gradient constraints).

**Reply by abarun22** [reply]
Hi Mark,
I was trying to think of a framework from the multi-scale perspective and wanted to clarify the following. 
1) Concerning the derivative of first piola stress with respect to the macro deformation gradient, i am interested in the following relation when visualized from a macro model.
![image](https://user-images.githubusercontent.com/40111479/157252624-f3a1492d-f56c-4a31-aeae-6ede0ffe59bb.png)
The variable _pk1_stress_hom is volume averaged and supposed to be transferred from the micro model. My concern is whether this quantity is already present in the volume averaged condition or i may need to do that explicitly in the micro-model before i transfer that to the macro. For example when you look at this line `_pk1_stress[_qp](_indices[_h].first, _indices[_h].second)` this does not seem to be a volume averaged variable, as it has been referenced for a quadrature point and not for a domain. 
2) I planned to compute this derivative from within the macro model as this seems the most optimal place to incorporate this material property. I guess just naming this variable as `_pk1_jacobian` and placing the calculations inside `computeQpPK1Stress()` will do the job.
3) I presume that Newton's method will make use of the jacobian that we derive. How different is PJFNK from Newton's in this case?
Kind regards,
Arun

**Reply by reverendbedford** [reply]
1. If you do things as I've suggested (use the system to impose the macroscale deformation) you will need to calculate the volume-averaged stress yourself.  You can use a postprocessor to average the "pk1" material property over the microscale domain.
2. I don't think so.  You will need to calculate the derivative in the microscale problems and then transfer the information to your macroscale problem.  
3. Assuming you get your Jacobian correct Newton's method will likely convergence quicker (fewer iterations) and is generally faster for "standard" one scale mechanics problems.  But your situation is rather different.  If you don't want to get your Jacobian correct than you might use PJFNK, which is less sensitive to problems with the Jacobian.

**Reply by abarun22** [reply]
Hi Mark,
I wanted to recall one of my earlier question and clarify that before i move on with further activities. I am still not very clear how different are the homogenized RVE stresses as against the original macro stresses, since the method is based entirely on constraining the stresses or strains through the construction of residual. The stresses may not vary much even when the deformation gradients are allowed to drive the RVE model. If these arguments are valid, could this be relieved by adding the macro deformation gradient to the deformation gradient computed by the system? The later one is computed by the gradient of the micro-fluctuation field against the original RVE configuration. In that case we could also use the macro stresses instead of macro deformation gradients to drive the RVE. Do you see if these modifications are reasonable and compatible with the system assumptions? 
Kind regards,
Arun

**Reply by reverendbedford** [reply]
That's what the system already does.  Please review the math in the [documentation](https://mooseframework.inl.gov/modules/tensor_mechanics/Homogenization.html).

The displacements used in solving for the stresses in the microscale problem are the superposition of the  macroscale displacements and the microscale displacements.  If you entirely impose deformation gradient constraints, as I've suggested, the macroscale displacements are affine with the macroscale deformation gradient (your "macro deformation gradient".  The microscale displacements (your "micro-flucation field") are periodic but otherwise unconstrained.

**Reply by abarun22** [reply]
Yes Mark, i see that the macro displacements are superposed over the micro-scale displacements while formulating the constraints. That's the key in coupling the macro model with the fine scale model and we might be able to see the influence of one on the other. I am currently working on developing a relevant transfer scheme for carrying the tensors between the scales. Do you see any other issues concerning this model to be going in to my work? If things are OK, i shall look further in to integrating the system with my framework.
Kind regards,
Arun
